### PR TITLE
Fix LGT_TO_GT to be a registered function

### PR DESCRIPTION
**Pull request by tpoterba**
Now supports 0-ploid and haploid as well as diploid.

### CODE REVIEW

File: hail/src/main/scala/is/hail/variant/Call.scala [hunk 1]
```diff
@@ -95,13 +95,18 @@ object Call extends Serializable {
 
   def alleleRepr(c: Call): Int = c >>> 3
 
-  def allelePair(c: Call): AllelePair = {
+  def allelePair(c: Call): Int = {
```
Comment (tpoterba) [inline hail/src/main/scala/is/hail/variant/Call.scala:98]: Changes here and in Genotype.scala are to remove the AllelePair value class, which made code generation much harder and also introduced unintentional allocation in at least once place.

File: hail/src/main/scala/is/hail/types/physical/stypes/concrete/SCanonicalCall.scala [hunk 2]
```diff
@@ -91,6 +93,46 @@ class SCanonicalCallSettable(val call: Settable[Int]) extends SCallValue with SS
           cb.append(Code._fatal[Unit](const("invalid ploidy: ").concat(p.toS)))))
     })
   }
+
+  def lgtToGT(cb: EmitCodeBuilder, localAlleles: SIndexableValue, errorID: Value[Int]): SCallCode = {
```
Comment (tpoterba) [inline hail/src/main/scala/is/hail/types/physical/stypes/concrete/SCanonicalCall.scala:97]: lgt_to_gt is a function that translates LGT (local gt) to GT (absolute gt) using the local alleles array. 

For instance, with LGT=0/2 and LA=[0,3,4], the GT is 0/4.

**Comment by johnc1231** [comment]
Read through and looks good. Just checking: this is just intended to be an internal registered function, not something for users to call, right?

File: hail/src/main/scala/is/hail/types/physical/stypes/concrete/SCanonicalCall.scala [hunk 3]
```diff
@@ -91,6 +93,46 @@ class SCanonicalCallSettable(val call: Settable[Int]) extends SCallValue with SS
...
+            checkAndTranslate(cb, allele), isPhased())
+          cb.assign(repr, newCall)
+        }, // ploidy 1
+        EmitCodeBuilder.scopedVoid(cb.emb) { cb =>
+          val allelePair = cb.newLocal[Int]("allelePair", Code.invokeScalaObject1[Int, Int](Call.getClass, "allelePairUnchecked", call))
+          val j = cb.newLocal[Int]("allele_j", Code.invokeScalaObject1[Int, Int](AllelePair.getClass, "j", allelePair))
+          val k = cb.newLocal[Int]("allele_k", Code.invokeScalaObject1[Int, Int](AllelePair.getClass, "k", allelePair))
+
+          cb.ifx(j >= localAlleles.loadLength(), cb._fatal("invalid lgt_to_gt: allele "))
+          val indexableLen = localAlleles.loadLength()
```
Comment (johnc1231) [inline hail/src/main/scala/is/hail/types/physical/stypes/concrete/SCanonicalCall.scala:125]: Unused?

**Review by johnc1231 — CHANGES_REQUESTED** [review]

**Comment by tpoterba** [comment]
> Read through and looks good. Just checking: this is just intended to be an internal registered function, not something for users to call, right?

Users may sometimes use this -- it was exposed through `hl.experimental` previously as part of the vcf combiner utilities, I just moved it to `hl.vds` (though it still exists at the old location for compatibility)

File: hail/src/main/scala/is/hail/types/physical/stypes/concrete/SCanonicalCall.scala [hunk 4]
```diff
@@ -91,6 +93,45 @@ class SCanonicalCallSettable(val call: Settable[Int]) extends SCallValue with SS
...
+      cb.ifx(av >= localAlleles.loadLength(),
+        cb._fatalWithError(errorID,
+          s"lgt_to_gt: found allele ", av.toS, ", but there are only ", localAlleles.loadLength().toS, " local alleles"))
+      localAlleles.loadElement(cb, av).get(cb, const("lgt_to_gt: found missing value in local alleles at index ").concat(av.toS), errorID = errorID)
+        .asInt.intCode(cb)
+    }
+
+    val repr = cb.newLocal[Int]("lgt_to_gt_repr")
+    cb += Code.switch(ploidy(),
+      EmitCodeBuilder.scopedVoid(cb.emb)(cb => cb._fatal(s"ploidy above 2 is not currently supported")),
```
Comment (johnc1231) [inline hail/src/main/scala/is/hail/types/physical/stypes/concrete/SCanonicalCall.scala:110]: Should be `fatalWithID` and

File: hail/src/main/scala/is/hail/types/physical/stypes/concrete/SCanonicalCall.scala [hunk 5]
```diff
@@ -91,6 +93,45 @@ class SCanonicalCallSettable(val call: Settable[Int]) extends SCallValue with SS
...
+          val newCall = Code.invokeScalaObject2[Int, Boolean, Int](Call1.getClass, "apply",
+            checkAndTranslate(cb, allele), isPhased())
+          cb.assign(repr, newCall)
+        }, // ploidy 1
+        EmitCodeBuilder.scopedVoid(cb.emb) { cb =>
+          val allelePair = cb.newLocal[Int]("allelePair", Code.invokeScalaObject1[Int, Int](Call.getClass, "allelePairUnchecked", call))
+          val j = cb.newLocal[Int]("allele_j", Code.invokeScalaObject1[Int, Int](AllelePair.getClass, "j", allelePair))
+          val k = cb.newLocal[Int]("allele_k", Code.invokeScalaObject1[Int, Int](AllelePair.getClass, "k", allelePair))
+
+          cb.ifx(j >= localAlleles.loadLength(), cb._fatal("invalid lgt_to_gt: allele "))
```
Comment (johnc1231) [inline hail/src/main/scala/is/hail/types/physical/stypes/concrete/SCanonicalCall.scala:124]: Should be `fatalWithId`

**Review by johnc1231 — DISMISSED** [review]
If users can call, we should pass `errorId`s thoroughly I think. Two places you are just using `fatal`

File: hail/src/main/scala/is/hail/types/physical/stypes/concrete/SCanonicalCall.scala (hunk 5 above)
Comment (tpoterba) [reply hail/src/main/scala/is/hail/types/physical/stypes/concrete/SCanonicalCall.scala:124]: good catch, thanks!

**Review by johnc1231 — APPROVED** [review]

**Pull request merged**